File: tree/114.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return 
        # pre-order traversal
        l = []
        l.append(root)
        while l:
            cur = l.pop()
            if cur.right:
                l.append(cur.right)
            if cur.left:
                l.append(cur.left)
            if l:
                cur.right = l[-1] #如果left 存在则指向left 否则指向right
            cur.left = None

File: tree/114.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
import java.util.List;
import java.util.ArrayList;

class Solution {
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        }
        TreeNode dummy = new TreeNode(-1);
        dummy.right = root;
        List<TreeNode> list = new ArrayList<>();
        if (root.right != null) {
            list.add(root.right);
        }
        if (root.left != null) {
            list.add(root.left);
        }
        while (list.size() > 0) {
            TreeNode cur = list.remove(list.size()-1);
            root.left = null;
            root.right = cur;
            root = root.right;
            if (cur.right != null) {
                list.add(cur.right);
            }
            if (cur.left != null) {
                list.add(cur.left);
            }
        }
        root = dummy.right;
    }
}
